### PR TITLE
add functionality to load Solcast API data to `iotools`

### LINKED ISSUE #1313: Integrate Solcast API into iotools
**Is your feature request related to a problem? Please describe.**
Get irradiance via Solcast API
https://solcast.com/solar-data-api/api/

**Describe the solution you'd like**
Use iotools to get irradiance data from Solcast

**Describe alternatives you've considered**
Solcast already has an API toolkit. Is irons redundant?

**Additional context**
* https://docs.solcast.com.au/#introduction
* https://solcast.com/solar-data-api/api-toolkit/
* icymi @AdamRJensen

### PR DESCRIPTION
**Pull request by lorenzo-solcast**
- [x] Closes #1313 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

With this PR we are proposing to add a module `iotools.solcast` with functions to get irradiation and weather data for live, forecast, historic and tmy from the Solcast API. We noticed the demand for such functionality and many of our users use PVLib with Solcast data.

Note that Solcast has a free tier for radiation and weather data which you can access immediately by using a company email address when [signing up](https://toolkit.solcast.com.au/register). We also have unlimited requests at [unmetered locations](https://docs.solcast.com.au/#unmetered-locations) and additional free data is also [available](https://solcast.com/data-for-researchers) for researchers and students.

This is a first iteration and are looking for feedback, in particular about the following design decisions:

1. following the pattern in the other `iotools` modules, we have created a mapping to translate variables form Solcast's definition to PVLibs that is applied when `map_variables` is `True`. This is a 1-way translation so the user would still query the Solcast API with Solcast's definitions and get the variables in a `pd.DataFrame` with the variables converted to PVLib's. We could make the interface only using PVLib's definitions but that may cause confusion with the API documentation. 
2. testing is currently using `mocks`. If you prefer we can provide an API key to test directly against the API.
3. the function signatures have only the required parameters declared with all optional as `kwargs`. The intention is to avoid docstrings that are too verbose.

Some tests have been added but we'll add more after a first review. 

Thanks!

### CODE REVIEW

**Comment by AdamRJensen** [comment]
@lorenzo-solcast thank you for this PR!

Note, we have not yet taken an official stance on iotools functions for commercial data providers so we can't promise this will get merged. Personally I'm in favor, under the condition that you take on the maintenance of the associated functions.

The irradiance iotools functions follow a pattern of returning a tuple of (data, meta). If there's no metadata to parse then just return an empty dictionary. Perhaps down the road metadata will be returned, and even if not then it's offered to have a standardized format.

Mock testing is probably sufficient (there's other iotools that only have that). Maybe it's even preferred.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:52]: Does this only happen when `map_variables` is True?

I think it would be preferable to always have the same time index convention, so I would move this comment to the output section.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:54]: This doesn't render correctly in the documentation. I would suggest making a references section and then linking to that,  e.g.,

    Optional parameters passed to the API. See [1]_ for full list of parameters.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:59]:
```suggestion
    data : pandas.DataFrame
        containing the values for the parameters requested

```

Should also output a meta dictionary, even if empty

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:63]: This also does not render correctly in the documentation. Code needs to be indented four spaces.

Also, I would prefer if you added the `pvlib.iotools` to the functions, making it easier for new users to copy the code examples.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:74]: pythons time zone convention is a bit counterintuitive (see the [this link](https://pvlib-python.readthedocs.io/en/stable/user_guide/timetimezones.html)) so I would recommend adding a code comment stating what 10 corresponds to.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:123]: Perhaps provide an example of what an ISO 8601 time duration is.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:283]:
```suggestion
    use ``map_variables``=False to avoid converting the data to pvlib's conventions

```

**Review by AdamRJensen — COMMENTED** [review]
An initial review from the airport😄

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: yes, it is applied only if `map_variables` is True. The idea being that the raw data is returned if set to `False`, but we can do that by default if preferred?

**Comment by lorenzo-solcast** [comment]
thanks for the review @AdamRJensen! should have addressed most of your points and returning an empty `dict` to start with.

> Note, we have not yet taken an official stance on iotools functions for commercial data providers so we can't promise this will > get merged. Personally I'm in favor, under the condition that you take on the maintenance of the associated functions.

Well aware and totally up to the team. We'd be happy to contribute and take on the maintenance of the module.

**Comment by lorenzo-solcast** [comment]
hi! any other thoughts on this PR?

**Review by cwhanse — APPROVED** [review]
Thanks @lorenzo-solcast 

@pvlib/pvlib-maintainer I support adding this.

**Comment by lorenzo-solcast** [comment]
thanks for the review @cwhanse - your suggestion makes the docstring much clearer.

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:71]: In all functions, the metadata dictionary should be listed in the `Returns` section as well.

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:114]: Is there no way to get metadata (data version, e.g.) back from the API?

Let's at least echo the `latitude` and `longitude` inputs in the metadata dict.

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:160]:
```suggestion
    kwargs:
        Optional parameters passed to the API.
        See [2]_ for full list of parameters.
```

For `kwargs`, we should be sure to be clear about the difference between GET parameters and parameters to `requests.get`.   Same for the other functions

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:77]:
```suggestion
    >>> df, meta = pvlib.iotools.solcast.get_solcast_tmy(
```

And same for the other examples

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:364]:
```suggestion
def _solcast2pvlib(data):
```

No need for this function to be public IMHO

File: pvlib/tests/iotools/test_solcast.py
Comment (kandersolar) [inline pvlib/tests/iotools/test_solcast.py:3]: We should test `_historical` and `_forecast` too, right?

**Comment by AdamRJensen** [comment]
@lorenzo-solcast If you're able to address these issues within the week, we should be able to get these functions into the December pvlib release 🥳

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: done.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: added reference

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py:71]: done.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: data version not yet! I have added the lat/lon as suggested.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: done.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: done.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py]: done.

File: pvlib/tests/iotools/test_solcast.py
Comment (lorenzo-solcast) [reply pvlib/tests/iotools/test_solcast.py]: done.

**Comment by lorenzo-solcast** [comment]
thanks for the great review @kandersolar! should have addressed all the points you raised - let me know what you think.

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:41]: @AdamRJensen This is an interesting approach that we might consider adopting in some centralized way for iotools in general.

No action needed in this PR, of course.

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:20]: A few more renames are needed, I think:
- clearsky_dhi -> dhi_clear
- clearsky_dni -> dni_clear
- clearsky_ghi -> ghi_clear
- clearsky_gti -> poa_global_clear (??)

@AdamRJensen any thoughts on `clearsky_gti`?

File: pvlib/iotools/solcast.py
Comment (kandersolar) [inline pvlib/iotools/solcast.py:67]: I think this should not be [3]

**Review by kandersolar — APPROVED** [review]
A couple more minor things but otherwise LGTM!

@AdamRJensen can you merge this when you think it's ready?

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [reply pvlib/iotools/solcast.py:20]: I suppose this is consistent with the usage of ``poa_global`` and ``_clear``. What are you thinking @kandersolar ?

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [reply pvlib/iotools/solcast.py:41]: This is pretty cool

File: pvlib/iotools/solcast.py
Comment (kandersolar) [reply pvlib/iotools/solcast.py:20]: Was mostly wondering if we already had a name for clear-sky GTI, and if not, looking for a second opinion that `poa_global_clear` is a good choice :)

**Comment by lorenzo-solcast** [comment]
@kandersolar thanks for spotting. Turns out that the Pandas version (0.25) used in that Python 3.7 workflow has an issue with parsing ISO8601 durations without days. I have implemented [a hacky fix](https://github.com/lorenzo-solcast/pvlib-python/blob/be27d850d4441d1a853ed1d7fab79233b9b467b8/pvlib/iotools/solcast.py#L387) that will work as we never provide monthly periods, but it is not ideal. 

Do you guys have an alternate/preferred/suggestion pattern to fix this?

**Comment by kandersolar** [comment]
> Do you guys have an alternate/preferred/suggestion pattern to fix this?

I don't think we have a solution for this already (no other part of pvlib deals with ISO 8601 duration strings, to my knowledge).  That hacky fix seems reasonable to me.  If you want, you could leave a comment next to it indicating the better/simpler code we could switch to when we eventually increase our minimum pandas to 1.2.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:270]: I think it would be nice to add forecast horizon to this example using the ``hours`` parameter.

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:135]:
```suggestion
    """Get historical irradiance and weather estimated actuals.
```
I'm unfamiliar with the term "actuals" but maybe that is standard? As a non-native, I can't really tell what meaning is trying to be conveyed

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:158]: I think we went away from listing "default is None". In either case I would be consistent with the ``end`` and ``duration`` parameters which both are optional/defaults to None. I would prefer removing ", default is None"

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:156]:
```suggestion
        ISO 8601 compliant duration for the historic data,
```

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:158]:
```suggestion
        Must be within 31 days of ``start``.
```

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:433]: Right now the documentation of the child functions ``map_variables`` parameter only states:
>  When true, renames columns of the DataFrame to pvlib variable names where applicable

I think that all instance of ``map_variables`` should also denote that the timestamp is shifted (I don't recall seeing this anywhere else in pvlib, but do support the idea).

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:228]:
```suggestion
    """Get irradiance and weather forecasts from the present time
    up to 14 days ahead.
```

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:298]:
```suggestion
    and past 7 days.
```

**Review by AdamRJensen — APPROVED** [review]
Solid code! Only minor stuff from me.

Note, I wasn't able to test examples of the forecast/live data though with my API key.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py:329]: @AdamRJensen we have just given your account the required permissions to query for unmetered locations  if you want to test again!

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py:116]: that is a valid point, and relevant for our SDK too. There are no use-cases for this functionality yet, but I agree that down the line we may need expose this to the user.

File: pvlib/iotools/solcast.py
Comment (lorenzo-solcast) [reply pvlib/iotools/solcast.py:135]: we refer to "live" and "historical" data as _estimated actuals_. These would be sometimes be called "measurements" I guess, but we avoid that term as we are really just estimating these values from satellite observations and not direct measurements of weather/power. Safe to assume that anyone using the Solcast API understands this?

Comment (AdamRJensen) [reply pvlib/iotools/solcast.py:137]: Hmm I think that's a very Solcast specific terminology. To be more in line with the other iotools I suggest removing mentions of "actuals" and just calling them "estimates"

File: pvlib/iotools/solcast.py
Comment (AdamRJensen) [inline pvlib/iotools/solcast.py:424]:
```suggestion
    """Retrieve weather, irradiance and power data from the Solcast API.
```

File: pvlib/tests/iotools/test_solcast.py
Comment (AdamRJensen) [inline pvlib/tests/iotools/test_solcast.py:2]:
```suggestion
from pvlib.iotools import (
```
I would import it as a user would

**Review by AdamRJensen — APPROVED** [review]
I think we should be good to merge this function later today 🥳 

Great work @lorenzo-solcast!

**Comment by AdamRJensen** [comment]
I would also recommend adding a "See Also" section to all the functions pointing to their Solcast cousins. Example:

    See Also
    --------
    pvlib.iotools.read_solaranywhere, pvlib.iotools.get_bsrn

**Comment by AdamRJensen** [comment]
@lorenzo-solcast I made some minor changes so don't forget to pull from GH

I see that there are two lines that are not being hit by the tests, could you make some basic tests to hit these?
image

**Comment by lorenzo-solcast** [comment]
@AdamRJensen added the tests!

**Comment by AdamRJensen** [comment]
@lorenzo-solcast Access to Solcast data will now be supported by pvlib within a matter of days with the release of v0.10.3. Very exciting stuff! Thanks for your contribution and patience during the review process 🥳

**Comment by lorenzo-solcast** [comment]
exciting indeed! looking forward to contribute further and thank you for the thorough reviews.

**Pull request merged**
